python-package/xgboost/testing/data.py

```diff
@@ -124,3 +124,58 @@ def pd_dtypes() -> Generator:
         orig = pd.DataFrame(data, dtype=np.bool_ if Null is None else pd.BooleanDtype())
         df = pd.DataFrame(data, dtype=pd.BooleanDtype())
         yield orig, df
+
+
+def pd_arrow_dtypes() -> Generator:
+    """Pandas DataFrame with pyarrow backed type."""
+    import pandas as pd
+    import pyarrow as pa
+
+    import xgboost
+
+    # Integer
+    dtypes = xgboost.data.pandas_pyarrow_mapper
+    Null: Union[float, None, Any] = np.nan
+    orig = pd.DataFrame(
+        {"f0": [1, 2, Null, 3], "f1": [4, 3, Null, 1]}, dtype=np.float32
+    )
+    # Create a dictionary-backed dataframe, enable this when the roundtrip is
+    # implemented in pandas/pyarrow
+    #
+    # category = pd.ArrowDtype(pa.dictionary(pa.int32(), pa.int32(), ordered=True))
+    # df = pd.DataFrame({"f0": [0, 2, Null, 3], "f1": [4, 3, Null, 1]}, dtype=category)
+
+    # Error:
+    # >>> df.astype("category")
+    #   Function 'dictionary_encode' has no kernel matching input types
+    #   (array[dictionary<values=int32, indices=int32, ordered=0>])
```

Comment:
This may have been an issue in an older pyarrow version? While debugging this call I'm getting

```
In [8]: df.astype("category")
> /Users/.../pandas/core/arrays/arrow/array.py(656)factorize()
-> null_encoding = "mask" if use_na_sentinel else "encode"
(Pdb) self
<ArrowExtensionArray>
[0, 2, <NA>, 3]
Length: 4, dtype: dictionary<values=int32, indices=int32, ordered=1>[pyarrow]
(Pdb) self._data
<pyarrow.lib.ChunkedArray object at 0x7f7c6a932a40>
[

  -- dictionary:
    [
      0,
      2,
      3
    ]
  -- indices:
    [
      0,
      1,
      null,
      2
    ]
]
(Pdb) n
> /Users/.../pandas/core/arrays/arrow/array.py(657)factorize()
-> encoded = self._data.dictionary_encode(null_encoding=null_encoding)
(Pdb) null_encoding
'mask'
(Pdb) self._data.dictionary_encode(null_encoding=null_encoding)
*** pyarrow.lib.ArrowNotImplementedError: Function 'dictionary_encode' has no kernel matching input types (dictionary<values=int32, indices=int32, ordered=1>)
```

Reply:
I'm using pyarrow 10.0.1 from pypi:
``` python
import numpy as np
import pandas as pd
import pyarrow as pa

print(pd.__version__)
print(pa.__version__)

Null = np.nan
category = pd.ArrowDtype(pa.dictionary(pa.int32(), pa.int32(), ordered=True))
df = pd.DataFrame({"f0": [0, 2, Null, 3], "f1": [4, 3, Null, 1]}, dtype=category)
df.astype("category")
```
```
1.5.2
10.0.1
Traceback (most recent call last):
  File "/home/jiamingy/Workspace/XGBoost-dev/XGBoostUtils/pyarrow/test.py", line 11, in <module>
    df.astype("category")
  File "/home/jiamingy/.anaconda/envs/xgboost_dev/lib/python3.10/site-packages/pandas/core/generic.py", line 6233, in astype
    results = [
  File "/home/jiamingy/.anaconda/envs/xgboost_dev/lib/python3.10/site-packages/pandas/core/generic.py", line 6234, in <listcomp>
    self.iloc[:, i].astype(dtype, copy=copy)
  File "/home/jiamingy/.anaconda/envs/xgboost_dev/lib/python3.10/site-packages/pandas/core/generic.py", line 6240, in astype
    new_data = self._mgr.astype(dtype=dtype, copy=copy, errors=errors)
  File "/home/jiamingy/.anaconda/envs/xgboost_dev/lib/python3.10/site-packages/pandas/core/internals/managers.py", line 450, in astype
    return self.apply("astype", dtype=dtype, copy=copy, errors=errors)
  File "/home/jiamingy/.anaconda/envs/xgboost_dev/lib/python3.10/site-packages/pandas/core/internals/managers.py", line 352, in apply
    applied = getattr(b, f)(**kwargs)
  File "/home/jiamingy/.anaconda/envs/xgboost_dev/lib/python3.10/site-packages/pandas/core/internals/blocks.py", line 526, in astype
    new_values = astype_array_safe(values, dtype, copy=copy, errors=errors)
  File "/home/jiamingy/.anaconda/envs/xgboost_dev/lib/python3.10/site-packages/pandas/core/dtypes/astype.py", line 299, in astype_array_safe
    new_values = astype_array(values, dtype, copy=copy)
  File "/home/jiamingy/.anaconda/envs/xgboost_dev/lib/python3.10/site-packages/pandas/core/dtypes/astype.py", line 227, in astype_array
    values = values.astype(dtype, copy=copy)
  File "/home/jiamingy/.anaconda/envs/xgboost_dev/lib/python3.10/site-packages/pandas/core/arrays/base.py", line 608, in astype
    return cls._from_sequence(self, dtype=dtype, copy=copy)
  File "/home/jiamingy/.anaconda/envs/xgboost_dev/lib/python3.10/site-packages/pandas/core/arrays/categorical.py", line 495, in _from_sequence
    return Categorical(scalars, dtype=dtype, copy=copy)
  File "/home/jiamingy/.anaconda/envs/xgboost_dev/lib/python3.10/site-packages/pandas/core/arrays/categorical.py", line 441, in __init__
    codes, categories = factorize(values, sort=True)
  File "/home/jiamingy/.anaconda/envs/xgboost_dev/lib/python3.10/site-packages/pandas/core/algorithms.py", line 785, in factorize
    codes, uniques = values.factorize(  # type: ignore[call-arg]
  File "/home/jiamingy/.anaconda/envs/xgboost_dev/lib/python3.10/site-packages/pandas/core/arrays/arrow/array.py", line 603, in factorize
    encoded = self._data.dictionary_encode(null_encoding=null_encoding)
  File "pyarrow/table.pxi", line 586, in pyarrow.lib.ChunkedArray.dictionary_encode
  File "pyarrow/_compute.pyx", line 560, in pyarrow._compute.call_function
  File "pyarrow/_compute.pyx", line 355, in pyarrow._compute.Function.call
  File "pyarrow/error.pxi", line 144, in pyarrow.lib.pyarrow_internal_check_status
  File "pyarrow/error.pxi", line 121, in pyarrow.lib.check_status
pyarrow.lib.ArrowNotImplementedError: Function 'dictionary_encode' has no kernel matching input types (dictionary<values=int32, indices=int32, ordered=0>)
```

Reply:
Ah okay. This might be a a bug in pandas `astype`ing in 1.5.2.

On the main branch (for pandas 2.0), I am getting ordered=1

```
In [1]: import numpy as np
   ...: import pandas as pd
   ...: import pyarrow as pa
   ...:
   ...: print(pd.__version__)
   ...: print(pa.__version__)
   ...:
   ...: Null = np.nan
   ...: category = pd.ArrowDtype(pa.dictionary(pa.int32(), pa.int32(), ordered=True))
   ...: df = pd.DataFrame({"f0": [0, 2, Null, 3], "f1": [4, 3, Null, 1]}, dtype=category)
2.0.0.dev0+1220.gd72e244db5
10.0.1

In [2]: df.astype("category")
---------------------------------------------------------------------------
ArrowNotImplementedError                  Traceback (most recent call last)
Cell In [2], line 1
----> 1 df.astype("category"
...
ArrowNotImplementedError: Function 'dictionary_encode' has no kernel matching input types (dictionary<values=int32, indices=int32, ordered=1>)
```

Reply:
Looking forward to the new pandas!